### PR TITLE
RAII FatFS minus the filesystem corruption

### PR DESCRIPTION
**Pull request by stellar-aria**
Let's try this again

### CODE REVIEW

**Comment by stellar-aria** [comment]
Same PR as before, this time with checked access using `.value()`

File: src/deluge/model/sample/sample_recorder.cpp [hunk 1]
```diff
@@ -696,8 +696,8 @@ Error SampleRecorder::finalizeRecordedFile() {
 				if (firstSampleCluster->sdAddress == 0) {
 					FREEZE_WITH_ERROR("E268");
 				}
-				if ((firstSampleCluster->sdAddress - fileSystemStuff.fileSystem->database)
-				    & (fileSystemStuff.fileSystem->csize - 1)) {
+				if ((firstSampleCluster->sdAddress - fileSystemStuff.fileSystem.value().database)
```
Comment (m-m-adams) [inline src/deluge/model/sample/sample_recorder.cpp:699]: I don't think calling .value() is safe either - https://en.cppreference.com/w/cpp/utility/optional/value

Docs say it throws if the optional isn't present

Comment (m-m-adams) [reply src/deluge/model/sample/sample_recorder.cpp:699]: This isn't UB (unlike ->) but will crash deluge

Comment (stellar-aria) [reply src/deluge/model/sample/sample_recorder.cpp]: if the filesystem isn't valid I'd rather it crash than try to corrupt the SDCard or something.

**Comment by stellar-aria** [comment]
I think it'll be better to merge this series as single commits so any bugs can be tracked down

**Review by m-m-adams — APPROVED** [review]

**Pull request merged**